lists-and-memory/problem-set-lists-and-memory.md

```diff
@@ -288,78 +288,138 @@ As the loop executes with each iteration you may add an element to the `big_word
-        if numbers[mid] > value:
-            high = mid - 1
-        elif numbers[mid] < value:
-            low = mid + 1
-        else:
-            return mid
-    
-    
-    if numbers[low] == value:
-        return low
-    
-    return None
-```
+Here are the tests:
 
-##### !end-question
+```python
+def test_sort_by_length_with_empty_string():
+    assert sort_by_length("") == []
 
-##### !options
 
-* O(1)
-* O(log n)
-* O(n)
-* O(n<sup>2</sup>)
+def test_sort_by_length():
+    assert sort_by_length("I love great awesome words") == [
+        "I", "love", "great", "words", "awesome"]
 
-##### !end-options
 
-##### !answer
+def test_sort_by_length_checks_smallest_word_last():
+    assert sort_by_length("love great awesome words I") == [
+        "I", "love", "great", "words", "awesome"]
 
-* O(log n)
 
-##### !end-answer
+def test_sort_by_length_equal_length_maintains_order():
+    assert sort_by_length("words of equal length") == [
+        "of", "words", "equal", "length"]
+```
 
-<!-- other optional sections -->
-<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
-##### !rubric
+### !end-question
+### !placeholder
 
-Notice that the loop divides in half with each iteration.
+```python
+def sort_by_length(sentence):
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
 
-##### !end-rubric
+class TestPython1(unittest.TestCase):
+    def test_sort_by_length_with_empty_string(self):
+        self.assertEqual(sort_by_length(""), [])
 
-##### !explanation
+    def test_sort_by_length(self):
+        self.assertEqual(sort_by_length("I love great awesome words"),
+                         ["I", "love", "great", "words", "awesome"])
 
-The loop will check the middle of the list and it will either return the item, if found, or change high or low, skipping the half of the list which cannot contain the element.  So with each iteration the loop bypasses 1/2 of the remaining elements.
+    def test_sort_by_length_checks_smallest_word_last(self):
+        self.assertEqual(sort_by_length("love great awesome words I"),
+                         ["I", "love", "great", "words", "awesome"])
 
-If the list was 16 elements long, we would be searching 16 elements, then 8, then 4, then 2, then 1 and we either find the element or return `None`.
+    def test_sort_by_length_equal_length_maintains_order(self):
+        self.assertEqual(sort_by_length("words of equal length"),
+                         ["of", "words", "equal", "length"])
+```
+### !end-tests
+### !explanation
 
-This algorithm is known as [binary search](https://www.geeksforgeeks.org/python-program-for-binary-search/) and you can read more about the algorithm by following the link.
+An example of a working implementation:
 
-##### !end-explanation
+```python
+def sort_by_length(sentence):
+    array = sentence.split()
+    i = 1
+    while i < len(array):
+        to_insert = array[i]
+        j = i
```

Comment:
I think `j` is a bit too terse of a name here. Additionally I think of `i` and `j` as paired names, usually going through a 2D array or something like that, but the pairing between `i` and `j` here doesn't seem that.

I think we can hint better at what purpose `j` serves.

`insertion_index` or `insert_where` is something that came to my mind, as we are trying to find the right place to insert `to_insert`.

Reply:
hrmm. I like this thought and I think it could make the students first sort algorithm a little more approachable. However, when I was looking around at implementations of sorts on the internet, `i` and `j` were very commonly used, so I wonder if using them here would be helpful. Wanted to bring it up for more discussion before making a decision. @kaidamasaki do you have any thoughts on this one? @chimerror do you feel strongly that `i` and `j` aren't the right choice.

Reply:
I do feel strongly, but that's because I think people tend to be a bit more terse than they should be, but not so strongly you should hold yourself up on it!

Reply:
Yeah, `i` and `j` are pretty common but I think for teaching more descriptive names are a good idea.

There is a school of thought that short lived variable should have short names but I think being clearer isn't a bad idea and is really helpful as documentation.

Reply:
Thanks for the additional thoughts! I changed `j` to `insertion_index` and confirmed the code works